worlds/jakanddaxter/Rules.py

```diff
@@ -0,0 +1,222 @@
+import typing
+from BaseClasses import CollectionState
+from Options import OptionError
+from . import JakAndDaxterWorld
+from .Options import (EnableOrbsanity,
+                      GlobalOrbsanityBundleSize,
+                      PerLevelOrbsanityBundleSize,
+                      FireCanyonCellCount,
+                      MountainPassCellCount,
+                      LavaTubeCellCount,
+                      CitizenOrbTradeAmount,
+                      OracleOrbTradeAmount)
+from .locs import CellLocations as Cells
+from .Locations import location_table
+from .Levels import level_table
+
+
+def set_orb_trade_rule(world: JakAndDaxterWorld):
+    options = world.options
+    player = world.player
+
+    if options.enable_orbsanity == EnableOrbsanity.option_off:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_vanilla(state, player, world, required_orbs, required_previous_trade))
+    else:
+        world.can_trade = lambda state, required_orbs, required_previous_trade: (
+            can_trade_orbsanity(state, player, world, required_orbs, required_previous_trade))
+
+
+def recalculate_reachable_orbs(state: CollectionState, player: int, world: JakAndDaxterWorld) -> None:
+
+    if not state.prog_items[player]["Reachable Orbs Fresh"]:
+
+        # Recalculate every level, every time the cache is stale, because you don't know
+        # when a specific bundle of orbs in one level may unlock access to another.
+        for level in level_table:
+            state.prog_items[player][f"{level} Reachable Orbs".strip()] = (
+                count_reachable_orbs_level(state, world, level))
+
+        # Also recalculate the global count, still used even when Orbsanity is Off.
+        state.prog_items[player]["Reachable Orbs"] = count_reachable_orbs_global(state, world)
+        state.prog_items[player]["Reachable Orbs Fresh"] = True
+
+
+def count_reachable_orbs_global(state: CollectionState,
+                                world: JakAndDaxterWorld) -> int:
+
+    accessible_orbs = 0
+    for level_regions in world.level_to_regions.values():
+        for region in level_regions:
+            # Rely on short-circuiting to skip region.can_reach whenever possible.
+            if region.orb_count > 0 and region.can_reach(state):
+                accessible_orbs += region.orb_count
+    return accessible_orbs
+
+
+def count_reachable_orbs_level(state: CollectionState,
+                               world: JakAndDaxterWorld,
+                               level_name: str = "") -> int:
+
+    accessible_orbs = 0
+    for region in world.level_to_regions[level_name]:
+        # Rely on short-circuiting to skip region.can_reach whenever possible.
+        if region.orb_count > 0 and region.can_reach(state):
```

Comment:
Rather than doing this, I think it would be nice if you built a `world.level_to_regions_which_have_orbs` where the 0 regions are just culled

Reply:
I'm gonna be pretty nitpicky about these functions because (without having looked at it) this is probably where most of the performance concerns are gonna be

Reply:
I have some ideas to get what you want, I just hope they don't cost too much elsewhere - the magician's secret often comes down to trading time for space.

I also hope I don't have to individually add regions to that dictionary, so I have a shortcut in mind.

Reply:
The calculation for your world right now is definitely that you can use like 100x more space to get this to be more performant :D

Reply:
Didn't realize we had that much leg room! How's the solution that I put together in `create_regions`? Acceptable, or not enough?

Reply:
Totally fine